contractcourt/channel_arbitrator.go

```diff
@@ -1269,15 +1340,254 @@ func (c *ChannelArbitrator) isPreimageAvailable(hash lntypes.Hash) (bool,
+	for htlcIndex, htlc := range remoteHTLCs {
+		if _, ok := localHTLCs[htlcIndex]; ok {
+			continue
+		}
+
+		pendingRemoteHTLCs = append(pendingRemoteHTLCs, htlc)
+	}
+
+	// Finally, we'll examine all the pending remote HTLCs for those that
+	// have expired. If we find any, then we'll recommend that they be
+	// failed now so we can free up the incoming HTLC.
+	for _, htlc := range pendingRemoteHTLCs {
+		// We'll now check if we need to go to chain in order to cancel
+		// the incoming HTLC.
+		goToChain := c.shouldGoOnChain(
+			htlc.RefundTimeout, c.cfg.OutgoingBroadcastDelta,
+			height,
+		)
+
+		// If we don't need to go to chain, and no commitments have
+		// been confirmed, then we can move on. Otherwise, if
+		// commitments have been confirmed, then we need to cancel back
+		// *all* of the pending remote HTLCS.
+		if !goToChain && !commitsConfirmed {
+			continue
+		}
+
+		log.Tracef("ChannelArbitrator(%v): immediately failing "+
+			"htlc=%x from remote commitment",
+			c.cfg.ChanPoint, htlc.RHash[:])
+
+		actionMap[HtlcFailNowAction] = append(
+			actionMap[HtlcFailNowAction], htlc,
+		)
+	}
+
+	return actionMap
+}
+
+// checkRemoteChainActions examines the two possible remote commitment chains
+// and returns the set of chain actions we need to carry out if the remote
+// commitment (non pending) confirms. The pendingConf indicates if the pending
+// remote commitment confirmed. This is similar to checkCommitChainActions, but
+// we'll immediately fail any HTLCs on the pending remote commit, but not the
+// remote commit (or the other way around).
+func (c *ChannelArbitrator) checkRemoteChainActions(
+	height uint32, trigger transitionTrigger,
+	activeHTLCs map[HtlcSetKey]htlcSet,
+	pendingConf bool) (ChainActionMap, error) {
+
+	// First, we'll examine all the normal chain actions on the remote
+	// commitment that confirmed.
+	confHTLCs := activeHTLCs[RemoteHtlcSet]
+	if pendingConf {
+		confHTLCs = activeHTLCs[RemotePendingHtlcSet]
+	}
+	remoteCommitActions, err := c.checkCommitChainActions(
+		height, trigger, confHTLCs,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// With this actions computed, we'll now check the diff of the HTLCs on
+	// the commitments, and cancel back any that are on the pending but not
+	// the non-pending.
+	remoteDiffActions := c.checkRemoteDiffActions(
+		height, activeHTLCs, pendingConf,
+	)
+
+	// Finally, we'll merge all the chain actions and the final set of
+	// chain actions.
+	remoteCommitActions.Merge(remoteDiffActions)
+	return remoteCommitActions, nil
+}
+
+// checkRemoteDiffActions checks the set difference of the HTLCs on the remote
+// confirmed commit and remote dangling commit for HTLCS that we need to cancel
+// back. If we find any HTLCs on the remote pending but not the remote, then
+// we'll mark them to be failed immediately.
+func (c *ChannelArbitrator) checkRemoteDiffActions(height uint32,
+	activeHTLCs map[HtlcSetKey]htlcSet,
+	pendingConf bool) ChainActionMap {
```

Comment:
I think it would be clearer to pass in the two set keys instead of the boolean. Make it a more generic diff "htlcs in B but not in A". Then what A and B are can be specified in `checkRemoteChainActions`.

Reply:
If we actually need to generalize it in the future, then it's a small diff away. Atm, we only care about pending or not.